File: tests/rosetta/transpiler/Clojure/church-numerals-1.clj
(ns main (:refer-clojure :exclude [zero succ add mul pow incr toInt intToChurch]))

(require 'clojure.set)

(defn in [x coll]
  (cond (string? coll) (clojure.string/includes? coll x) (map? coll) (contains? coll x) (sequential? coll) (some (fn [e] (= e x)) coll) :else false))

(defn padStart [s w p]
  (loop [out (str s)] (if (< (count out) w) (recur (str p out)) out)))

(def nowSeed (atom (let [s (System/getenv "MOCHI_NOW_SEED")] (if (and s (not (= s ""))) (Integer/parseInt s) 0))))

(declare zero succ add mul pow incr toInt intToChurch)

(declare main_four main_three main_z pow_di pow_i pow_prod)

(defn zero [zero_f]
  (try (throw (ex-info "return" {:v (fn [x] (throw (ex-info "return" {:v x})))})) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn succ [succ_c]
  (try (throw (ex-info "return" {:v (fn [f] (throw (ex-info "return" {:v (fn [x] (throw (ex-info "return" {:v (f ((c f) x))})))})))})) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn add [add_c add_d]
  (try (throw (ex-info "return" {:v (fn [f] (throw (ex-info "return" {:v (fn [x] (throw (ex-info "return" {:v ((c f) ((d f) x))})))})))})) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn mul [mul_c mul_d]
  (try (throw (ex-info "return" {:v (fn [f] (throw (ex-info "return" {:v (fn [x] (throw (ex-info "return" {:v ((c (d f)) x)})))})))})) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn pow [pow_c pow_d]
  (try (do (def pow_di (toInt pow_d)) (def pow_prod pow_c) (def pow_i 1) (while (< pow_i pow_di) (do (def pow_prod (mul pow_prod pow_c)) (def pow_i (+ pow_i 1)))) (throw (ex-info "return" {:v pow_prod}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn incr [incr_i]
  (try (throw (ex-info "return" {:v (+ (int incr_i) 1)})) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn toInt [toInt_c]
  (try (throw (ex-info "return" {:v (int ((c incr) 0))})) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn intToChurch [intToChurch_i]
  (try (if (= intToChurch_i 0) zero (succ (intToChurch (- intToChurch_i 1)))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(def main_z zero)

(def main_three (succ (succ (succ main_z))))

(def main_four (succ main_three))

(defn -main []
  (let [rt (Runtime/getRuntime)
    start-mem (- (.totalMemory rt) (.freeMemory rt))
    start (System/nanoTime)]
      (println (str "three        -> " (str (toInt main_three))))
      (println (str "four         -> " (str (toInt main_four))))
      (println (str "three + four -> " (str (toInt (add main_three main_four)))))
      (println (str "three * four -> " (str (toInt (mul main_three main_four)))))
      (println (str "three ^ four -> " (str (toInt (pow main_three main_four)))))
      (println (str "four ^ three -> " (str (toInt (pow main_four main_three)))))
      (println (str "5 -> five    -> " (str (toInt (intToChurch 5)))))
      (System/gc)
      (let [end (System/nanoTime)
        end-mem (- (.totalMemory rt) (.freeMemory rt))
        duration-us (quot (- end start) 1000)
        memory-bytes (Math/abs ^long (- end-mem start-mem))]
        (println (str "{\n  \"duration_us\": " duration-us ",\n  \"memory_bytes\": " memory-bytes ",\n  \"name\": \"main\"\n}"))
      )
    ))

(-main)

File: tests/rosetta/transpiler/php/church-numerals-1.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
$__start_mem = memory_get_usage();
$__start = _now();
  function zero($f) {
  global $z, $three, $four;
  return function($x) use ($f) {
  return $x;
};
};
  function succ($c) {
  global $z, $three, $four;
  return function($f) use ($c) {
  return function($x) use ($c, $f, $call_user_func) {
  return $f(call_user_func($c($f), $x));
};
};
};
  function add($c, $d) {
  global $z, $three, $four;
  return function($f) use ($c, $d) {
  return function($x) use ($c, $d, $f, $call_user_func) {
  return call_user_func($c($f), call_user_func($d($f), $x));
};
};
};
  function mul($c, $d) {
  global $z, $three, $four;
  return function($f) use ($d, $c) {
  return function($x) use ($c, $d, $f, $call_user_func) {
  return call_user_func($c($d($f)), $x);
};
};
};
  function mochi_pow($c, $d) {
  global $z, $three, $four;
  $di = toInt($d);
  $prod = $c;
  $i = 1;
  while ($i < $di) {
  $prod = mul($prod, $c);
  $i = $i + 1;
};
  return $prod;
};
  function incr($i) {
  global $z, $three, $four;
  return (intval($i)) + 1;
};
  function toInt($c) {
  global $z, $three, $four;
  return intval(call_user_func($c('incr'), 0));
};
  function intToChurch($i) {
  global $z, $three, $four;
  if ($i == 0) {
  return 'zero';
}
  return succ(intToChurch($i - 1));
};
  $z = 'zero';
  $three = succ(succ(succ($z)));
  $four = succ($three);
  echo rtrim('three        -> ' . _str(toInt($three))), PHP_EOL;
  echo rtrim('four         -> ' . _str(toInt($four))), PHP_EOL;
  echo rtrim('three + four -> ' . _str(toInt(add($three, $four)))), PHP_EOL;
  echo rtrim('three * four -> ' . _str(toInt(mul($three, $four)))), PHP_EOL;
  echo rtrim('three ^ four -> ' . _str(toInt(mochi_pow($three, $four)))), PHP_EOL;
  echo rtrim('four ^ three -> ' . _str(toInt(mochi_pow($four, $three)))), PHP_EOL;
  echo rtrim('5 -> five    -> ' . _str(toInt(intToChurch(5)))), PHP_EOL;
$__end = _now();
$__end_mem = memory_get_usage();
$__duration = intdiv($__end - $__start, 1000);
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;;
